### PR TITLE
ActionCable Messages

### PR DESCRIPTION
**Pull request by navro**
**Changes:**
- Adds a "live chat" for messages using ActionCable/Websockets. This means a lot of code had to be moved from controllers/views to javascript and action cable files.
- Changes the design for messages - removes a lot of padding to make it look more like a chat
- Removes `MessageComment` - I don't think we need them
- Updates config files to add a redis database for action cable
- Adds `ejs` gem to allow javascript templating
- Adds `puma` to `development` and `testing` since websockets don't work with `webrick`
- Adds push notifications for messages

**Next PR:**
- Add an API controller for messages
- Make token authorization work with action cable. This may be a bit tricky since Websockets doesn't support extra headers, but I think i have found a [solution
](https://gist.github.com/navro/886f7e6ce6b2110d9cc393d38e9b6172#file-gistfile1-txt)

**Comments:**
- It should be easy to add this to the app, since a lot of the code is written in javascript. The ActionCable client can be added to the app using npm.
- ActionCable has no built-in feature to track connected users. We need to have this since we don't want people to get push notifications when they are active on the chat page. My solution was to write a `ConnectedList` class that stores the list of connected users in redis.
- It was fun to do this, quite different from the regular ruby/rails programming we do 😄

### CODE REVIEW

File: config/initializers/action_cable_config.rb [hunk 1]
```diff
@@ -0,0 +1,11 @@
+module ActionCableConfig
+  # Read a value from cable.yml
+  def self.[](key)
+    unless @config
+      template = ERB.new(File.read(Rails.root + 'config/cable.yml'))
+      @config = YAML.load(template.result(binding))[Rails.env].symbolize_keys
```
Comment (navro) [inline config/initializers/action_cable_config.rb]: No. It doesn't work with "safe_load".

File: app/assets/javascripts/channels/groups.js [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+  var page = 2;
+  var totalPages = 37;
+  var loadingMessages = false;
+
+  var prependMessage = function(message) {
+    messages.prepend(template({message: message, current_user: userId}));
+  };
+
+  var loadMessages = function() {
```
Comment (navro) [inline app/assets/javascripts/channels/groups.js]: It's called directly after this definition. Are you stupid??

File: app/assets/javascripts/cable.js [hunk 3]
```diff
@@ -0,0 +1,10 @@
+//= require action_cable
+//= require_self
+//= require_tree ./channels
+
+(function() {
+  this.App || (this.App = {});
```
Comment (navro) [inline app/assets/javascripts/cable.js]: I don't care, this is a file automatically created by rails.

File: app/channels/application_cable/channel.rb [hunk 4]
```diff
@@ -0,0 +1,6 @@
+module ApplicationCable
```
Comment (navro) [inline app/channels/application_cable/channel.rb]: Auto generated by rails, I won't fix this

File: config/initializers/sidekiq.rb [hunk 5]
```diff
@@ -1,4 +1,9 @@
-redis = { namespace: Rails.env, url: "redis://localhost:6379/#{ENV['REDIS_DB_NUMBER'] || 0}" }
+redis = { namespace: Rails.env, url: "redis://localhost:6379/#{ENV['SIDEKIQ_DB_NUMBER'] || 0}" }
+
+if Rails.env == "test"
+  redis = { namespace: Rails.env, url: "redis://localhost:6379/0" }
```
Comment (navro) [inline config/initializers/sidekiq.rb]: And where do you see those?

File: app/assets/javascripts/channels/groups.js [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+    $.getJSON('/grupper/' + groupId + '/meddelanden/')
+      .then(function(resp) {
+        for (var message of resp.messages) prependMessage(message);
+        totalPages = resp.meta.total_pages;
```
Comment (navro) [inline app/assets/javascripts/channels/groups.js]: Variables that come from the json data created in rails. Not going to edit that for some JS guidelines

File: app/assets/javascripts/channels/groups.js [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+  var appendMessage = function(message) {
+    messages.append(template({message: message, current_user: userId}));
+    to_bottom();
+  };
+
+  var removeMessage = function(messageId) {
+    $('.message#' + messageId).remove();
+  };
+
+  var updateMessage = function(message) {
```
Comment (navro) [inline app/assets/javascripts/channels/groups.js]: Wrong

File: app/channels/application_cable/connection.rb [hunk 8]
```diff
@@ -0,0 +1,21 @@
+module ApplicationCable
```
Comment (navro) [inline app/channels/application_cable/connection.rb]: Auto generated by rails, I won't fix this

File: app/assets/javascripts/channels/groups.js [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+    $.getJSON('/grupper/' + groupId + '/meddelanden/')
+      .then(function(resp) {
+        for (var message of resp.messages) prependMessage(message);
+        totalPages = resp.meta.total_pages;
+        toBottom();
+      });
+  }();
+
+  var appendMessage = function(message) {
+    messages.append(template({message: message, current_user: userId}));
```
Comment (navro) [inline app/assets/javascripts/channels/groups.js]: Variables that come from the json data created in rails. Not going to edit that for some JS guidelines

File: app/assets/javascripts/channels/groups.js [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+  App.groups = App.cable.subscriptions.create({
+    channel: "GroupsChannel",
+    group_id: groupId },
+  { connected: function() {},
+    disconnected: function() {},
+    received: function(data) {
+      if (data.action == 'create') {
+        appendMessage(data.message.message);
+      } else if (data.action == 'destroy') {
+        removeMessage(data.message_id);
```
Comment (navro) [inline app/assets/javascripts/channels/groups.js]: Variables that come from the json data created in rails. Not going to edit that for some JS guidelines

File: app/assets/javascripts/channels/groups.js [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+    $('.message#' + messageId).remove();
+  };
+
+  var updateMessage = function(message) {
+    // If the message is loaded, update it!
+    var msg = $('.message#' + message.id);
+    if(!msg.length) return;
+
+    msg.find('.msg-content').html(message.content);
+    msg.find('.updated').html(message.updated_at);
```
Comment (navro) [inline app/assets/javascripts/channels/groups.js]: Variables that come from the json data created in rails. Not going to edit that for some JS guidelines

File: app/assets/javascripts/channels/groups.js [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+      } else if (data['action'] == 'destroy') {
+        removeMessage(data['message_id']);
+      } else if (data['action'] == 'update') {
+        updateMessage(data['message'].message);
+      }
+    },
+    sendMessage: function(content) {
+      return this.perform('send_message', {
+        content: content,
+        group_id: groupId
```
Comment (navro) [inline app/assets/javascripts/channels/groups.js]: Variables that come from the json data created in rails. Not going to edit that for some JS guidelines

File: app/assets/javascripts/channels/groups.js [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+      });
+    },
+    destroyMessage: function(messageId) {
+      return this.perform('destroy_message', {
+        message_id: messageId
```
Comment (navro) [inline app/assets/javascripts/channels/groups.js]: Variables that come from the json data created in rails. Not going to edit that for some JS guidelines

File: config/initializers/sidekiq.rb (hunk 5 above)
Comment (davidwessman) [inline config/initializers/sidekiq.rb]: Check for [hard spaces](https://en.wikipedia.org/wiki/Hard_space), probably after the first `{`.
I write them all the time.

Comment (navro) [inline config/initializers/sidekiq.rb]: +1
You were right!

Comment (davidwessman) [inline config/initializers/sidekiq.rb]: It was probably me who wrote it ^^

**Pull request merged**